Bigtable/tests/Unit/DataClientTest.php

```diff
@@ -0,0 +1,240 @@
+    }
+
+    public function testMutateRowsFailure()
+    {
+        $statuses = [];
+        $status = new Status;
+        $status->setCode(Code::INVALID_ARGUMENT);
+        $status->setMessage('Invalid argument');
+        $statuses[] = $status;
+        $status = new Status;
+        $status->setCode(Code::OK);
+        $statuses[] = $status;
+
+        $mutateRowsResponses = $this->getMutateRowsResponse($statuses);
+        $this->serverStream->readAll()
+            ->shouldBeCalled()
+            ->willReturn(
+                $this->arrayAsGenerator($mutateRowsResponses)
+            );
+        $this->bigtableClient->mutateRows(self::TABLE_NAME, $this->entries, $this->options)
+            ->shouldBeCalled()
+            ->willReturn(
+                $this->serverStream->reveal()
+            );
+        try {
+            $this->dataClient->mutateRows($this->rowMutations);
+            $this->fail('Expected exception is not thrown');
+        } catch (BigtableDataOperationException $e) {
+            $metadata = [
+                [
+                    'rowKey' => 'rk1',
+                    'rowMutationIndex' => 0,
+                    'statusCode' => Code::INVALID_ARGUMENT,
+                    'message' => 'Invalid argument'
+                ]
+            ];
+            $this->assertEquals('Invalid argument', $e->getMessage());
+            $this->assertEquals(Code::INVALID_ARGUMENT, $e->getCode());
+            $this->assertEquals(
+                $metadata,
+                $e->getMetadata()
+            );
+        }
+    }
+
+    /**
+     * @expectedException Google\ApiCore\ApiException
+     * @expectedExceptionMessage unauthenticated
+     */
+    public function testMutateRowsApiExceptionInMutateRows()
+    {
+        $apiException =  new ApiException('unauthenticated', Code::UNAUTHENTICATED, 'unauthenticated');
+        $this->bigtableClient->mutateRows(self::TABLE_NAME, $this->entries, $this->options)
+            ->shouldBeCalled()
+            ->willThrow(
+                $apiException
+            );
+        $this->dataClient->mutateRows($this->rowMutations);
+    }
+
+    /**
+     * @expectedException Google\ApiCore\ApiException
+     * @expectedExceptionMessage unauthenticated
+     */
+    public function testMutateRowsApiExceptionInReadAll()
+    {
+        $apiException =  new ApiException('unauthenticated', Code::UNAUTHENTICATED, 'unauthenticated');
+        $this->serverStream->readAll()
+            ->shouldBeCalled()
+            ->willThrow(
+                $apiException
+            );
+        $this->bigtableClient->mutateRows(self::TABLE_NAME, $this->entries, $this->options)
+            ->shouldBeCalled()
+            ->willReturn(
+                $this->serverStream->reveal()
+            );
+        $this->dataClient->mutateRows($this->rowMutations);
+    }
+
+    public function testUpsert()
+    {
+        $statuses = [];
+        foreach ($this->rowMutations as $rowMutation) {
```

Comment:
$rowMutation is never used?
If you want to create an array of many status objects, perhaps just for loop until count($this->rowMutations) will be more straightforward.